---
title: "Filtering and multiple sequence alignment of FLC-like proteins"
author: "Edward Wallace"
date: "2022-07-08"
output: 
  html_document:
    toc:
      true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(janitor)
library(Biostrings)
library(ggmsa)
library(taxize)
```

# Summary

This document is for filtering and multiple sequence alignment of FLC-like proteins.

It takes:

- protein sequences from `data/uniprot/uniprot-proteins-IPR040241-20220708.fasta`
- list of representative taxa from XXXXX

Then:

- filters the sequences only from those taxa TO DO
- multiple sequence alignment of chosen sequences TO DO
- maybe, make a nice figure of the multiple sequence alignment? TO DO
- trim the alignments TO DO
- maybe, a quick approximate phylogenetic tree using fasttree? TO DO

# Load data

## Load protein metadata

```{r load_protein_metadata}
protein_metadata <- 
  here::here("data", "uniprot", "uniprot-proteins-IPR040241-20220708.tsv") %>%
  readr::read_tsv() %>%
  janitor::clean_names(case = "big_camel")

protein_metadata
```

## Load protein sequences

```{r load_protein_sequences}
protein_seqs <- 
  here::here("data", "uniprot", "uniprot-proteins-IPR040241-20220708.fasta") %>%
  Biostrings::readAAStringSet() %>%
  magrittr::set_names(
    stringr::word(names(.), start = 2L, sep = fixed("|"))
  )

protein_seqs
```

## Load taxonIDs and get taxon lineage dataframe

```{r load_taxonIDs}
taxon_IDs <- 
  here::here("data", "uniprot", "TaxonIDs.txt") %>%
  readr::read_tsv()

taxon_IDs

taxon_lineage <- taxize::classification(
  taxon_IDs$TaxonID, 
  db='ncbi') %>%
  cbind() %>% 
  select(c("clade", "phylum", "subphylum", "class","order","family","genus", "species", "strain", "query"))%>% 
  mutate_at(10, as.numeric) %>%
  dplyr::rename(taxonID = query) 

taxon_lineage
```

# Filter data 

## Filter uniprot metadata by taxonIDs from curated list of species

```{r filter_uniprot_metadata_by_taxonID}

filtered_protein_metadata <- 
  filter(protein_metadata, OrganismId %in% taxon_IDs$TaxonID) %>% 
    dplyr::rename(uniprotID = Entry, taxonID = OrganismId, length = Length) %>% 
    left_join(taxon_lineage) %>% 
    select(c("uniprotID", "clade", "phylum", "subphylum", "class","order","family","genus", "species", "strain", "taxonID", "length"))
    

filtered_protein_metadata %>% 
  dplyr::filter(phylum == "Chytridiomycota")

```

## Filter uniprot fasta sequences by uniprotIDs from curated species and write new fasta file

```{r filter_uniprot_fasta_by_taxonID}

filtered_protein_fasta <-
protein_seqs[filtered_protein_metadata$uniprotID]

filtered_protein_fasta

writeXStringSet(filtered_protein_fasta, here::here("data", "MAFFT-alignment", "filtered-uniprot-proteins-IPR040241-20220708.fasta"))
```




## View length of proteins in the data set 

I notcied that some of the proteins in the data set are much smalled than S. cerevisiae Flc2, so I made a quick histogram of protien length. I think that protiens <400 amino acids are probably misannotated or do not have the transmembrane domains. 

```{r fig.width=4, fig.height=3}
protein_length_histogram <-
  ggplot(filtered_protein_metadata, aes(x=length)) + 
  geom_histogram(binwidth=50, color="darkblue", fill="lightblue") + 
  theme_classic()

protein_length_histogram

small_proteins <- filtered_protein_metadata %>%
  filter(length < 400)

small_proteins$uniprotID
```
# View topology of protien in the dataset 

I predicted the topology of the proteins in my filtered list using DeepTMHMM. I ran the following code from the terminal, where PATH = the local file path to filtered_protein_fasta. 

biolib run DTU/DeepTMHMM --fasta PATH 

biolib run DTU/DeepTMHMM --fasta /Users/s1427471/Documents/PhD/FLC-like-phylogeny/data/MAFFT-alignment/filtered-uniprot-proteins-IPR040241-20220708.fasta

I saved the gff3 file in "data/protein_topology/DeepTMHMM_prediction/Deep_TMHMM_filtered-uniprot-proteins-IPR040241-20220708.gff3" and used the following code look at the transmembrane topology. 

```{r}
DeepTMHMM_result <- 
read_table(
  here::here("data/protein_topology/DeepTMHMM_prediction/Deep_TMHMM_filtered-uniprot-proteins-IPR040241-2022-07-20.gff3"), 
  col_names = c("uniprotID", "topology","dom_start","dom_end"),
  col_types = "ccii",
  comment = "#"
  ) %>% 
  mutate(
    topology = factor(topology, 
      levels = c("signal", "outside", "inside","periplasm", "TMhelix", "Beta sheet"),
      labels = c("signal peptide", "outside", "inside", "periplasm","transmembrane helix", "beta sheet"))
  ) %>% 
  
  full_join(filtered_protein_metadata)

summary(DeepTMHMM_result)

```

# Plot number of TM helices histogram 

This chunk counts the number of TM helices in each protien and plots a histogram. It also outputs a list of protiens with no TM helices because these are not included in the histogram. Note that most proteins have nine transmembrane helices. 

```{r}

TM_helix_df <- DeepTMHMM_result %>% 
  group_by(uniprotID) %>% 
  dplyr::filter(topology == "transmembrane helix") %>% 
  mutate(topology = as.character(topology)) %>% 
  count(topology) %>% 
  dplyr::rename(TM_helix_number = n)


TM_helix_number_histogram <- 
  ggplot(TM_helix_df, aes(x = TM_helix_number)) +
  geom_histogram(binwidth=1, color="darkblue", fill="lightblue") + 
  theme_classic()
  
TM_helix_number_histogram

# This chunk prints proteins with no transmembrane domains 
a <- unique(DeepTMHMM_result$uniprotID)
b <- unique(TM_helix_df$uniprotID)
no_TMs <- union(setdiff(a,b), setdiff(b,a))
no_TMs

```

# View predict protein topology of all FLC/PKD2 domain protien from select species 

This chuck plots the predicted transmembrane topology of each protein with an IPR040241 domain found in our select list of species. Note that some proteins have 
```{r fig.height=5, fig.width=13}

myfourcolours <- c("#e07a5f","#f2cc8f","#81b29a", "#6a4c93")

seg_size <- 2
big_text <- 15
small_text <- 12
my_theme <-  theme(
    axis.title = element_text(size = big_text),
    legend.title = element_text(size = big_text),
    legend.text = element_text(size = small_text)
  )

topology_plot <- 
  ggplot(data = DeepTMHMM_result, aes(x = dom_start, xend = dom_end, y = uniprotID, yend = uniprotID, colour = topology)) +
  geom_segment(size =seg_size, lineend = "square") + 
  theme_classic() +
facet_grid(phylum ~ . , scales = "free_y", space = "free", switch = "y") +
  labs(x = "residue", y = "phylum") +
  theme(strip.placement = "outside", 
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0, hjust=1),
        axis.title.y = element_blank(),
        axis.text.y  = element_blank())

topology_plot + scale_color_manual(values = myfourcolours)+ 
geom_text(data = DeepTMHMM_result %>% group_by(uniprotID) %>% 
filter(dom_start == min(dom_start)),
aes(x = dom_start, y = uniprotID, label = uniprotID),
nudge_x = -10,
hjust = 1,
show.legend = FALSE, 
colour = "black",
size = 2
)
#ggsave(here::here("data/output_images/topology_plot_all_protiens_from_select_taxa.png"), height = 12, width = 15)


```

# Select protiens with nine transmembrane alpha helcies 
 
This chunk further filters out protein list to contain sequences with nine predicted transmembrane alpha helices and plots their predicted topology.

```{r}

nineTM_protiens <- TM_helix_df %>% 
  filter(TM_helix_number == 9)

nineTM_protien_topology <- DeepTMHMM_result %>% 
  filter(uniprotID %in% nineTM_protiens$uniprotID) %>% 
  group_by(uniprotID)

nineTM_topology_plot <- 
  ggplot(data = nineTM_protien_topology, aes(x = dom_start, xend = dom_end, y = uniprotID, yend = uniprotID, colour = topology)) +
  geom_segment(size =seg_size, lineend = "square") + 
  theme_classic() +
facet_grid(phylum ~ . , scales = "free_y", space = "free", switch = "y") +
  labs(x = "residue", y = "phylum") +
  theme(strip.placement = "outside", 
        strip.background = element_blank(),
        strip.text.y.left = element_text(angle = 0, hjust=1),
        axis.title.y = element_blank(),
        axis.text.y  = element_blank())

nineTM_topology_plot + scale_color_manual(values = myfourcolours)+ 
geom_text(data = nineTM_protien_topology %>% group_by(uniprotID) %>% 
filter(dom_start == min(dom_start)),
aes(x = dom_start, y = uniprotID, label = uniprotID),
nudge_x = -10,
hjust = 1,
show.legend = FALSE, 
colour = "black",
size = 2
)

#ggsave(here::here("data/output_images/topology_plot_9TM_protiens_from_select_taxa.png"), height = 12, width = 15)

```


# Truncate nine TM protiens by domain 

This chuck take our list of 9TM proteins and writes a new fasta file. It also writes fasta files containing the pre- trans membrane N-terminal sequences, the transmembrane region sequences and the post-transmembrane C-terminal sequences. 

```{r}
nineTM_protein_fasta <-
protein_seqs[nineTM_protiens$uniprotID]
nineTM_protein_fasta
writeXStringSet(filtered_protein_fasta, here::here("data", "MAFFT-alignment", "9TM_full_length.fasta"))





```



