---
title: "Filtering and multiple sequence alignment of FLC-like proteins"
author: "Edward Wallace"
date: "2022-07-08"
output: 
  html_document:
    toc:
      true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(janitor)
library(Biostrings)
```

# Summary

This document is for filtering and multiple sequence alignment of FLC-like proteins.

It takes:

- protein sequences from `data/uniprot/uniprot-proteins-IPR040241-20220708.fasta`
- list of representative taxa from XXXXX

Then:

- filters the sequences only from those taxa TO DO
- multiple sequence alignment of chosen sequences TO DO
- maybe, make a nice figure of the multiple sequence alignment? TO DO
- trim the alignments TO DO
- maybe, a quick approximate phylogenetic tree using fasttree? TO DO

# Load data

## Load protein metadata

```{r load_protein_metadata}
protein_metadata <- 
  here::here("data", "uniprot", "uniprot-proteins-IPR040241-20220708.tsv") %>%
  readr::read_tsv() %>%
  janitor::clean_names(case = "big_camel")

protein_metadata
```

## Load protein sequences

```{r load_protein_sequences}
protein_seqs <- 
  here::here("data", "uniprot", "uniprot-proteins-IPR040241-20220708.fasta") %>%
  Biostrings::readAAStringSet() %>%
  magrittr::set_names(
    stringr::word(names(.), start = 2L, sep = fixed("|"))
  )

protein_seqs
```

## Load taxonIDs 

```{r load_taxonIDs}
taxon_IDs <- 
  here::here("data", "uniprot", "TaxonIDs.txt") %>%
  readr::read_tsv()

taxon_IDs
```

# Filter data 

## Filter uniprot metadata by taxonIDs from curated list of species

```{r filter_uniprot_metadata_by_taxonID}

filtered_protein_metadata <- 
  filter(protein_metadata, OrganismId %in% taxon_IDs$TaxonID)

filtered_protein_metadata
```

## Filter uniprot fasta sequences by uniprotIDs from curated species and write new fasta file

```{r filter_uniprot_fasta_by_taxonID}

filtered_protein_fasta <-
protein_seqs[filtered_protein_metadata$Entry]

filtered_protein_fasta

writeXStringSet(filtered_protein_fasta, here::here("data", "MAFFT-alignment", "filtered-uniprot-proteins-IPR040241-20220708.fasta"))
```



This is the command I ran from the terminal to get a MAFFT alignment, but I don't know how to incorporate it into an .rmd file - TO DO 

command=
"/usr/local/bin/mafft"  --localpair  --maxiterate 16 --reorder "data/MAFFT-alignment/filtered-uniprot-proteins-IPR040241-20220708.fasta" > "data/MAFFT-alignment/MAFFT-aligned.fasta"





